lti_tool/views.py

```diff
@@ -0,0 +1,317 @@
+        resource = DeepLinkResource()
+        (resource.set_url(request.build_absolute_uri(reverse('lti-launch')))
+            .set_custom_params({
+                "course_slug": kwargs['course_slug'],
+                "instance_slug": kwargs['instance_slug']
+            })
+            .set_title("{}: {}".format(self.course.name, self.instance.instance_name)))
+        return HttpResponse(deep_link.output_response_form([resource]))
+
+    # Get list of modules in course
+    def get(self, request, *args, **kwargs):
+        self.modules = list(self.instance.course_modules.all())
+        self.note("modules")
+        return super().get(request, *args, **kwargs)
+    
+
+class LtiSelectModuleView(CourseModuleBaseView, LtiSelectContentMixin):
+
+    template_name = "lti_tool/exercise_list.html"
+
+    def post(self, request, *args, **kwargs):
+        # Create gradebook items
+        ags = self.message_launch.get_ags()
+        lineitems = ags.get_lineitems()
+        exercises = BaseExercise.objects.filter(course_module=self.module)
+        for e in exercises:
+            lineitem_exists = bool([li for li in lineitems if li['tag'] == str(e.id)])
+            if e.max_points > 0 and not lineitem_exists:
+                li = LineItem()
+                li.set_tag(str(e.id)).set_score_maximum(e.max_points).set_label(str(e))
+                ags.find_or_create_lineitem(li)
+
+        # Send activity settings
+        deep_link = self.message_launch.get_deep_link()
+        resource = DeepLinkResource()
+        (resource.set_url(request.build_absolute_uri(reverse('lti-launch')))
+            .set_custom_params({
+                "course_slug": kwargs['course_slug'],
+                "instance_slug": kwargs['instance_slug'],
+                "module_slug": kwargs['module_slug']
+            })
+            .set_title(str(self.module)))
+        return HttpResponse(deep_link.output_response_form([resource]))
+
+    # Get list of exercises in module
+    def get(self, request, *args, **kwargs):
+        self.learningobjects = LearningObject.objects.filter(course_module=self.module)
+        self.learningobjects_dict = { obj.id: obj for obj in self.learningobjects }
+        self.flat_module = self.content.flat_module(self.module)
+        exercises = [entry for entry in self.content.flat_module(self.module) if entry['type'] == 'exercise']
+        for exercise in exercises:
+            learningobj = self.learningobjects_dict[exercise['id']]
+            exercise['link'] = learningobj.get_url('lti-select-exercise')
+        self.note("learningobjects", "flat_module")
+        return super().get(request, *args, **kwargs)
+
+
+class LtiSelectExerciseView(ExerciseBaseView, LtiSelectContentMixin):
+
+    def post(self, request, *args, **kwargs):
+        # Create gradebook items
+        ags = self.message_launch.get_ags()
+        children = self.exercise.children.all()
+        if len(children) > 0:
```

Comment:
Does the existence of children imply that this is a chapter with embedded exercises in it? Then, the gradebook items are created for the child exercises, right?

This is not important now, but in theory, any learning object may have several layers of children. I think `exercise.children` includes only the direct children, not the "grandchildren" etc. We don't really have any courses that use such extra layers of children.

Reply:
Yea, right now it only creates gradebook items for children and not grandchildren - was not aware there could be more layers.